### ISSUE TITLE
Upgrade nightmare

### ISSUE THREAD

**Issue by hmeerlo** [aws/karpenter-provider-aws#6056]
Hi, I love the functionality of Karpenter but upgrading has been a complete nightmare until now. I have version `v0.16.3` running on several clusters. But I now have to upgrade EKS because it will not be supported anymore anytime soon (v1.24 now). I have tried to upgrade on 2 test clusters now but it is very confusing what the proper way to do this is. 

I first went to `v0.19.0` but then it crashed constantly because of missing settings (clusterName). I added that setting and it did something again. Then a few versions later it started complaining about missing permissions for the role which weren't described in the upgrade docs (`DescribeSpotPriceHistory` and `DescribeInstanceTypeOfferings`). Then somewhere along the line the new CRD's were introduced. But how do I know which version of these to use with which version of karpenter? And then finally I went to v0.34.0 and then it started complaining about missing field `cluster-name`.

Is there a documented path for going from an old version (0.16.3 in my case) to a recent version?

**Comment by jamessthompson** [comment]
Have you seen this? https://karpenter.sh/preview/upgrading/upgrade-guide/

**Comment by hmeerlo** [comment]
Yes I have seen that, I tried to follow that but ran into so many issues that I gave up. I assume it is easy when you're deep down into this codebase and exactly understand how things work. But for some devops guy that just needs to upgrade an installed Helm chart this is way too complex. Simple example from the Getting Started guide:

In [1.](https://karpenter.sh/docs/getting-started/getting-started-with-karpenter/#1-install-utilities): 

> Karpenter requires cloud provider permissions to provision nodes, for AWS IAM Roles for Service Accounts (IRSA) should be used. IRSA permits Karpenter (within the cluster) to make privileged requests to AWS (as the cloud provider) via a ServiceAccount.


Then further down:

> Karpenter now supports using [Pod Identity](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html) to authenticate AWS SDK to make API requests to AWS services using AWS Identity and Access Management (IAM) permissions. This feature not supported on versions of Kubernetes < 1.24. If you are running an earlier version of Kubernetes, you will need to use the [IAM Roles for Service Accounts(IRSA)](https://docs.aws.amazon.com/emr/latest/EMR-on-EKS-DevelopmentGuide/setting-up-enable-IAM.html) for pod authentication instead.

This makes me wonder what I should do during the upgrade?
What version of the CRD should I apply? Does it matter?
At some version the `clusterEndpoint` became required as a setting, then later on it disappeared again?
And somewhere the permissions for the controller apparently changed. How and during which version upgrade should I have applied what change for this?

**Comment by ClaytonOlleyNutrien** [comment]
I'm having similar difficulties going from 0.31 to 0.36 and all the fun CRD/API changes... I just gave up and completely removed the old version and installed the new, but even that is proving to be a nightmare with Terraform cycle errors which I believe are related to tags.

**Comment by jonathan-innis** [comment]
> I'm having similar difficulties going from 0.31 to 0.36 and all the fun CRD/API changes

@ClaytonOlleyNutrien What issues did you run into as you went through the upgrade process? Is there anything that we could add into the docs, upgrade guide, or the v1beta1 migration guide that could have made this process easier? Or was some of this due to translating the migration requirements from the docs over to TF?

**Comment by jonathan-innis** [comment]
> This makes me wonder what I should do during the upgrade

I agree that we should have called this out in the upgrade guide and that this was a miss when we added support for Pod Identity. This technically got added support in v0.32.2, but I'm going to add it to the "Upgrading to 0.33.x+" section since this is the first major version that contains this change.

> What version of the CRD should I apply? Does it matter

You should always match the CRD version that you are applying with the Karpenter application deployment's version. There's an expectation that these versions are coupled. I'd agree that this could be better-clarified in the docs.

> At some version the clusterEndpoint became required as a setting, then later on it disappeared again

This setting still exists and can be optionally specified. You can see [here](https://karpenter.sh/docs/reference/settings/#:~:text=controller%E2%80%99s%20TLS%20configuration.-,CLUSTER_ENDPOINT,-%2D%2Dcluster%2Dendpoint) that not specifying cluster endpoint allows Karpenter to auto-discover it now. If you don't want to give Karpenter `eks:DescribeCluster` permission, then you can just directly specify it through the settings. 

> And somewhere the permissions for the controller apparently changed. How and during which version upgrade should I have applied what change for this

The permissions got scoped down in v0.29.0 because they were quite broad and we wanted to make sure that our default "Getting Started" recommendation was still secure. I can make sure this gets back-ported back through the upgrade guide as with the other changes. 

If there's anything else that you think that we missed and has made the upgrade process frustrating, please feel free to share and help contribute to the upgrade guide so we can make the upgrade experience better for everyone.

**Comment by hmeerlo** [comment]
I have been thinking about this a lot lately, because I have to upgrade a few production systems soon. I will try to document my experience as good as I can. A few things I remember:

- the docs never even mention the actual command to execute to perform the upgrade. It just talks about new versions and their features and incompatibilities. 
- the repo changed for the artifacts. But it doesn't mention what to do when transitioning to the new repo. So in this case as well: what command do I need to perform to go from 0.16.3 to 0.17?
- do you have to upgrade through every single version, or can you skip versions? If so, which?
- 0.19.0: environment variables are removed and put into configMap. Should I do this manually? Does this happen automatically during the upgrade?
- 0.19.0: Not sure but I think it was this version where Karpenter started complaining about a missing clusterName and after clusterEndpoint. Apparently these weren't set before, but how could I have known that?

I will try to document as much as I can during my journey. But tbh I'm really nervous for performing these steps on a production system. I would really love for Karpenter to release a 1.0.0 version so we can start trusting on semver for these situations.

**Comment by hmeerlo** [comment]
Ok, I finished the production upgrades. I stopped at version `0.22.1` because that was the bare minimum I needed to support Kubernetes `>1.25`.
I ran into some issues but once I understood what to do, I could easily replicate the upgrade for the other clusters. Biggest thing I didn't realise was that while upgrading to `0.19.3` I had to supply the settings with `--set settings.aws.<setting?`. I expected it to be done automatically during the upgrade because the values were already in set in the environment. So besides all earlier mentioned issues, this was the only new one.

**Comment by hmeerlo** [comment]
Oh one final thing. It is not really cool that the docs for older versions are not online anymore. I'm now at `0.22.1` because of aforementioned issues, but without docs.
